File: pert5/tugas/mbti.pl
% --------------------------
% DATABASE
% --------------------------
:- dynamic sifat_pos/1.
:- dynamic sifat_neg/1.

% Daftar kepribadian
kepribadian("Singa").
kepribadian("Lumba-lumba").
kepribadian("Burung Hantu").
kepribadian("Semut").
kepribadian("Kura-kura").

% Relasi sifat dengan kepribadian
sifat(nyaman_keramaian, "Singa").
sifat(suka_merencanakan, "Singa").
sifat(hal_baru, "Singa").
sifat(tak_sesuai_rencana, "Singa").
sifat(bosan, "Singa").

sifat(nyaman_keramaian, "Lumba-lumba").
sifat(cemas, "Lumba-lumba").
sifat(hal_baru, "Lumba-lumba").
sifat(paham_perasaan, "Lumba-lumba").
sifat(bosan, "Lumba-lumba").
sifat(hindari_konflik, "Lumba-lumba").

sifat(suka_merencanakan, "Burung Hantu").
sifat(cemas, "Burung Hantu").
sifat(paham_perasaan, "Burung Hantu").
sifat(sendiri, "Burung Hantu").
sifat(tak_sesuai_rencana, "Burung Hantu").
sifat(hindari_konflik, "Burung Hantu").
sifat(sulit_percaya_orang, "Burung Hantu").

sifat(suka_merencanakan, "Semut").
sifat(sendiri, "Semut").
sifat(tak_sesuai_rencana, "Semut").
sifat(sulit_percaya_orang, "Semut").

sifat(suka_merencanakan, "Kura-kura").
sifat(cemas, "Kura-kura").
sifat(paham_perasaan, "Kura-kura").
sifat(sendiri, "Kura-kura").
sifat(tak_sesuai_rencana, "Kura-kura").
sifat(hindari_konflik, "Kura-kura").
sifat(sulit_percaya_orang, "Kura-kura").

% Pertanyaan untuk setiap sifat
pertanyaan(nyaman_keramaian, "Apakah kamu merasa nyaman di keramaian atau bertemu orang baru?").
pertanyaan(suka_merencanakan, "Apakah kamu suka merencanakan sesuatu jauh-jauh hari?").
pertanyaan(cemas, "Apakah kamu cenderung merasa cemas atau khawatir tanpa ada alasan yang jelas?").
pertanyaan(hal_baru, "Apakah kamu senang mencoba hal-hal baru yang belum pernah kamu lakukan sebelumnya?").
pertanyaan(paham_perasaan, "Apakah kamu mudah memahami perasaan orang lain tanpa harus dijelaskan?").
pertanyaan(sendiri, "Apakah kamu lebih suka bekerja sendiri daripada dalam kelompok?").
pertanyaan(tak_sesuai_rencana, "Apakah kamu merasa sangat terganggu jika sesuatu berjalan tidak sesuai dengan rencana?").
pertanyaan(bosan, "Apakah kamu sering merasa bosan jika rutinitas tidak berubah?").
pertanyaan(hindari_konflik, "Apakah kamu sering menghindari konflik atau konfrontasi langsung?").
pertanyaan(sulit_percaya_orang, "Apakah kamu merasa sulit mempercayai orang lain pada awal perkenalan?").

% --------------------------
% INFERENSI & INTERAKSI
% --------------------------

% Reset jawaban sebelum memulai
reset_jawaban :- 
    retractall(sifat_pos(_)),
    retractall(sifat_neg(_)).

% Menampilkan semua pertanyaan ke pengguna
tanya_semua :-
    forall(pertanyaan(ID, Teks), (
        format("~w (y/n): ", [Teks]),
        read(Jawaban),
        (Jawaban == y -> assertz(sifat_pos(ID)) ; assertz(sifat_neg(ID)))
    )).

% Menentukan hasil akhir
evaluasi_kepribadian(TipeTerbaik) :-
    findall(Tipe, kepribadian(Tipe), SemuaTipe),
    hitung_skor(SemuaTipe, SkorList),
    pilih_tertinggi(SkorList, TipeTerbaik).

% Hitung jumlah sifat yang cocok untuk setiap kepribadian
hitung_skor([], []).
hitung_skor([Tipe|Rest], [(Tipe,Skor)|SkorList]) :-
    findall(Sifat, (sifat(Sifat, Tipe), sifat_pos(Sifat)), SifatCocok),
    length(SifatCocok, Skor),
    hitung_skor(Rest, SkorList).

% Pilih kepribadian dengan skor tertinggi
pilih_tertinggi([(Tipe, Skor)], Tipe) :- !.
pilih_tertinggi([(Tipe1, Skor1), (Tipe2, Skor2)|Rest], TipeTerbaik) :-
    (Skor1 >= Skor2 ->
        pilih_tertinggi([(Tipe1, Skor1)|Rest], TipeTerbaik)
    ;
        pilih_tertinggi([(Tipe2, Skor2)|Rest], TipeTerbaik)
    ).

# % Program utama
# mulai :-
#     reset_jawaban,
#     writeln("=== Tes Kepribadian Hewan ==="),
#     tanya_semua,
#     evaluasi_kepribadian(Hasil),
#     format("~nKepribadian kamu adalah: ~w~n", [Hasil]).

File: pert5/tugas/mbti.py
from tkinter import *
from tkinter import messagebox
from pyswip import Prolog

# Pertanyaan sesuai ID Prolog
pertanyaan_dict = {
    "nyaman_keramaian": "Apakah kamu merasa nyaman di keramaian atau bertemu orang baru?",
    "suka_merencanakan": "Apakah kamu suka merencanakan sesuatu jauh-jauh hari?",
    "cemas": "Apakah kamu cenderung merasa cemas atau khawatir tanpa ada alasan yang jelas?",
    "hal_baru": "Apakah kamu senang mencoba hal-hal baru yang belum pernah kamu lakukan sebelumnya?",
    "paham_perasaan": "Apakah kamu mudah memahami perasaan orang lain tanpa harus dijelaskan?",
    "sendiri": "Apakah kamu lebih suka bekerja sendiri daripada dalam kelompok?",
    "tak_sesuai_rencana": "Apakah kamu merasa sangat terganggu jika sesuatu berjalan tidak sesuai dengan rencana?",
    "bosan": "Apakah kamu sering merasa bosan jika rutinitas tidak berubah?",
    "hindari_konflik": "Apakah kamu sering menghindari konflik atau konfrontasi langsung?",
    "sulit_percaya_orang": "Apakah kamu merasa sulit mempercayai orang lain pada awal perkenalan?"
}

#Deskripsi Tipe Kepribadian
deskripsi_kepribadian = {
    "Singa": """🦁 Sang Pemimpin (Singa)
Dominan, percaya diri, fokus pada tujuan. 
Kekuatan: Tegas, ambisius. 
Tantangan: Keras kepala, otoriter.""",

    "Lumba-lumba": """🐬 Sang Penjelajah (Lumba-lumba)
Kreatif, ekspresif, penuh empati. 
Kekuatan: Fleksibel, menyenangkan. 
Tantangan: Mudah terdistraksi, moody.""",

    "Burung Hantu": """🦉 Sang Pengamat (Burung Hantu)
Bijaksana, analitis, reflektif. 
Kekuatan: Mendalam, cermat. 
Tantangan: Cemas, terlalu banyak berpikir.""",

    "Semut": """🐜 Sang Pekerja Keras (Semut)
Tekun, fokus, efisien. 
Kekuatan: Mandiri, disiplin. 
Tantangan: Kurang fleksibel, kaku.""",

    "Kura-kura": """🐢 Sang Penjaga Kedamaian (Kura-kura)
Pendiam, penyabar, pengamat. 
Kekuatan: Penuh empati, setia. 
Tantangan: Menutup diri, menghindar dari tantangan."""
}

# Inisialisasi Prolog
prolog = Prolog()
prolog.consult("embeti.pl")

class KepribadianApp:
    def __init__(self, master):
        self.master = master
        self.master.title("Tes Kepribadian Hewan")
        self.index = 0
        self.sifat_ids = list(pertanyaan_dict.keys())
        self.jawaban_user = {}

        self.label = Label(master, text="", wraplength=400, font=("Helvetica", 12))
        self.label.pack(pady=20)

        self.btn_frame = Frame(master)
        self.btn_frame.pack(pady=10)

        self.yes_btn = Button(self.btn_frame, text="Ya", width=10, command=lambda: self.jawab("y"))
        self.yes_btn.grid(row=0, column=0, padx=10)

        self.no_btn = Button(self.btn_frame, text="Tidak", width=10, command=lambda: self.jawab("n"))
        self.no_btn.grid(row=0, column=1, padx=10)

        self.tampilkan_pertanyaan()

    def tampilkan_pertanyaan(self):
        if self.index < len(self.sifat_ids):
            sifat_id = self.sifat_ids[self.index]
            pertanyaan = pertanyaan_dict[sifat_id]
            self.label.config(text=pertanyaan)
        else:
            self.selesaikan_tes()

    def jawab(self, pilihan):
        sifat_id = self.sifat_ids[self.index]
        self.jawaban_user[sifat_id] = pilihan
        self.index += 1
        self.tampilkan_pertanyaan()

    def selesaikan_tes(self):
        prolog.query("reset_jawaban")
        for sifat, jawaban in self.jawaban_user.items():
            if jawaban == "y":
                prolog.assertz(f"sifat_pos({sifat})")
            else:
                prolog.assertz(f"sifat_neg({sifat})")

        hasil = list(prolog.query("evaluasi_kepribadian(X)"))
        if hasil:
            tipe = hasil[0]["X"].decode() if isinstance(hasil[0]["X"], bytes) else hasil[0]["X"]
            deskripsi = deskripsi_kepribadian.get(tipe, "Deskripsi Tidak Tersedia.")
            messagebox.showinfo("Hasil Kepribadian", f"Kepribadian kamu adalah: {tipe}\n\n{deskripsi}")
        else:
            messagebox.showinfo("Hasil", "Tidak dapat menentukan kepribadian.")

        self.master.destroy()

# Jalankan Aplikasi
if __name__ == "__main__":
    root = Tk()
    root.geometry("500x250")
    app = KepribadianApp(root)
    root.mainloop()
